src/main/java/io/supertokens/storage/mysql/Start.java

```diff
@@ -898,4 +902,305 @@ public void setJWTSigningKey_Transaction(TransactionConnection con, JWTSigningKe
+            throws StorageQueryException {
+        Connection sqlCon = (Connection) con.getConnection();
+        try {
+            PasswordlessQueries.deleteDevicesByEmail_Transaction(this, sqlCon, email);
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public void createCode(PasswordlessCode code) throws StorageQueryException, UnknownDeviceIdHash,
+            DuplicateCodeIdException, DuplicateLinkCodeHashException {
+
+        try {
+            PasswordlessQueries.createCode(this, code);
+        } catch (StorageTransactionLogicException e) {
+            // TODO: review these exceptions
+            if (e.actualException instanceof UnknownDeviceIdHash) {
+                throw (UnknownDeviceIdHash) e.actualException;
+            }
+            String message = e.actualException.getMessage();
+            if (message.matches(
+                    ".*Duplicate.*entry.*" + Config.getConfig(this).getPasswordlessCodesTable() + ".*PRIMARY\\'")) {
+                throw new DuplicateCodeIdException();
+            }
+            if (message.matches(".*Duplicate.*entry.*" + Config.getConfig(this).getPasswordlessCodesTable()
+                    + ".*link\\_code\\_hash\\'"))
+                if (message.contains("link_code_hash")) {
+                    throw new DuplicateLinkCodeHashException();
+                }
+            throw new StorageQueryException(e.actualException);
+        }
+    }
+
+    @Override
+    public PasswordlessCode getCodeByLinkCodeHash_Transaction(TransactionConnection con, String linkCodeHash)
+            throws StorageQueryException {
+        Connection sqlCon = (Connection) con.getConnection();
+        try {
+            return PasswordlessQueries.getCodeByLinkCodeHash_Transaction(this, sqlCon, linkCodeHash);
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public void deleteCode_Transaction(TransactionConnection con, String deviceIdHash) throws StorageQueryException {
+        Connection sqlCon = (Connection) con.getConnection();
+        try {
+            PasswordlessQueries.deleteCode_Transaction(this, sqlCon, deviceIdHash);
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public void createUser(io.supertokens.pluginInterface.passwordless.UserInfo user) throws StorageQueryException,
+            DuplicateEmailException, DuplicatePhoneNumberException, DuplicateUserIdException {
+        try {
+            PasswordlessQueries.createUser(this, user);
+        } catch (StorageTransactionLogicException e) {
+            // TODO: review these exceptions
+            String message = e.actualException.getMessage();
+            if (message.matches(
+                    ".*Duplicate.*entry.*" + Config.getConfig(this).getPasswordlessUsersTable() + ".*PRIMARY\\'") ||
```

Comment:
I like the change to regex, but I think they need a bit of cleanup: they allow different things and there are some strange things (escaped "_")

Reply:
Thank you, I love regular expressions. Sure, happy to discuss any changes. Generally symbols are escaped in regular expressions, to avoid any confusion this regex contains escaped underscores.

Reply:
Hmm, to be honest, seeing the escape sequence was more confusing than seeing just the underscore as that has no special meaning, so I was unsure if I forgot about something or if this was in some weird extension. :)